server/src/main/java/com/thoughtworks/go/server/service/ElasticAgentPluginService.java

```diff
@@ -193,68 +201,70 @@ public void createAgentsFor(List<JobPlan> old, List<JobPlan> newPlan) {
+                boolean secretsResolved = resolveSecrets(jobIdentifier, clusterProfile, elasticProfile);
+                if (secretsResolved) {
+                    createAgentQueue.post(new CreateAgentMessage(ephemeralAutoRegisterKeyService.autoRegisterKey(), environment, elasticProfile, clusterProfile, jobIdentifier), messageTimeToLive);
+                    serverHealthService.removeByScope(scopeForJob(jobIdentifier));
+                }
             } else {
-                String jobConfigIdentifier = plan.getIdentifier().jobConfigIdentifier().toString();
+                String jobConfigIdentifier = jobIdentifier.jobConfigIdentifier().toString();
                 String description = format("Plugin [%s] associated with %s is missing. Either the plugin is not " +
                         "installed or could not be registered. Please check plugins tab " +
                         "and server logs for more details.", clusterProfile.getPluginId(), jobConfigIdentifier);
-                serverHealthService.update(ServerHealthState.error(format("Unable to find agent for %s",
-                        jobConfigIdentifier), description, HealthStateType.general(HealthStateScope.forJob(plan.getIdentifier().getPipelineName(), plan.getIdentifier().getStageName(), plan.getIdentifier().getBuildName()))));
+                serverHealthService.update(error(format("Unable to find agent for %s",
+                        jobConfigIdentifier), description, general(scopeForJob(jobIdentifier))));
                 LOGGER.error(description);
             }
         }
     }
 
-    private void logToJobConsole(JobIdentifier identifier, String message) {
-        try {
-            consoleService.appendToConsoleLog(identifier, message);
-        } catch (IllegalArtifactLocationException | IOException e) {
-            LOGGER.error(format("Failed to add message(%s) to the job(%s) console", message, identifier), e);
-        }
-    }
-
-    private Predicate<JobPlan> isElasticAgent() {
-        return JobPlan::requiresElasticAgent;
-    }
-
     public boolean shouldAssignWork(ElasticAgentMetadata metadata, String environment, ElasticProfile elasticProfile, ClusterProfile clusterProfile, JobIdentifier identifier) {
-        Map<String, String> clusterProfileProperties = clusterProfile != null ? clusterProfile.getConfigurationAsMap(true) : Collections.EMPTY_MAP;
-        GoPluginDescriptor pluginDescriptor = pluginManager.getPluginDescriptorFor(metadata.elasticPluginId());
-        Map<String, String> configuration = elasticProfile.getConfigurationAsMap(true);
+        if (clusterProfile == null || !StringUtils.equals(clusterProfile.getPluginId(), metadata.elasticPluginId())) {
+            return false;
+        }
 
-        if (!StringUtils.equals(clusterProfile.getPluginId(), metadata.elasticPluginId())) {
+        boolean secretsResolved = resolveSecrets(identifier, clusterProfile, elasticProfile);
```

Comment:
It seems wrong for the `shouldAssignWork` method cancelling the job here. I think the caller of this method should handle the secret resolution exceptions and take necessary actions.

Reply:
I don't get it. If the service can cancel the job in the `createAgent` call, why can't it do so in `shouldAssignWork`?

Reply:
Hmmm, I think in both the cases I would have preferred the ElasticAgentPluginService not cancelling the job, the way create `createAgent` is implemented it makes it tough to make those changes.